test/base/strict-mode.js

```diff
@@ -241,6 +244,147 @@ test("warn on unsafe reads", function() {
+            })
+
+            const dispose = mobx.autorun(() => x.y)
+
+            utils.consoleWarn(() => {
+                x.y
+            }, /being read outside a reactive context/)
+
+            dispose()
+        } finally {
+            mobx.configure({ observableRequiresReaction: false })
+        }
+    })
+
+    test("warn on unsafe reads of observable array", function() {
+        try {
+            mobx.configure({ observableRequiresReaction: true })
+            const x = mobx.observable({
+                arr: [1, 2, 3]
+            })
+            utils.consoleWarn(() => {
+                x.arr[1]
+            }, /being read outside a reactive context/)
+        } finally {
+            mobx.configure({ observableRequiresReaction: false })
+        }
+    })
+    test("don't warn on reads inside batch - computed", function() {
+        try {
+            mobx.configure({ observableRequiresReaction: true })
+            const x = mobx.observable({
+                y: 1
+            })
+
+            const fooComputed = mobx.computed(() => x.y + 1)
+
+            const messages = utils.supressConsole(() => {
+                const dispose = mobx.autorun(() => fooComputed.get())
+                dispose()
+            })
+
+            expect(messages.length).toBe(0)
+        } finally {
+            mobx.configure({ observableRequiresReaction: false })
+        }
+    })
+
+    test("don't warn on reads inside batch - action", function() {
+        try {
+            mobx.configure({ observableRequiresReaction: true })
+            const x = mobx.observable({
+                y: 1
+            })
+
+            const fooAction = mobx.action(() => x.y)
+
+            const messages = utils.supressConsole(() => {
+                fooAction()
+            })
+
+            expect(messages.length).toBe(0)
+        } finally {
+            mobx.configure({ observableRequiresReaction: false })
+        }
+    })
+
+    test("don't warn on reads inside batch - transaction", function() {
+        try {
+            mobx.configure({ observableRequiresReaction: true })
+            const x = mobx.observable({
+                y: 1
+            })
+
+            const messages = utils.supressConsole(() => {
+                mobx.transaction(() => x.y)
+            })
+
+            expect(messages.length).toBe(0)
+        } finally {
+            mobx.configure({ observableRequiresReaction: false })
+        }
+    })
+
+    test("don't warn on reads inside batch - untracked inside autorun", function() {
+        try {
+            mobx.configure({ observableRequiresReaction: true })
+            const x = mobx.observable({
+                y: 1
+            })
+
+            const messages = utils.supressConsole(() => {
+                const disposer = mobx.autorun(() => {
+                    mobx.untracked(() => x.y)
+                })
+
+                disposer()
+            })
+
+            expect(messages.length).toBe(0)
```

Comment:
I don't understand why this passes....
The `untracked` sets derivation to null:
https://github.com/mobxjs/mobx/blob/251bb743860bc3f20809cc802a8330abf278dd68/src/core/derivation.ts#L294
and there is no batch ... therefore `x.y` should warn ... correct?

Reply:
But its still inside a batch, i've checked the code.

Reply:
Reaction uses startBatch inside,
And untracked only removed the `trackingDerivation`, not batch

https://github.com/mobxjs/mobx/blob/f944b5413f2f4b9de89e00e7de1ddde2a178e19d/src/core/reaction.ts#L92-L116

https://github.com/mobxjs/mobx/blob/f944b5413f2f4b9de89e00e7de1ddde2a178e19d/src/core/derivation.ts#L271-L279

Reply:
Ah right ... all reactions are inside batch ...

Reply:
Ok, so currently it doesn't warn because there is a batch ... but it shouldn't warn for different reason - existing derivation (or actually nonexsiting in case of `untracked`).
The goal is to force `x.y` being accessed either inside `reaction/autorun` or inside `action/tx`... right?
Is there a derivation which doesn't run inside batch so that `untracked` would warn?
Or similary ... is there a batch without `action/tx` so that `x.y` wouldn't warn?

Reply:
`enforceActions` uses `globalState.allowStateChanges`, shouldn't we use it instead of `inBatch` to detect allowed access outside derivation?

Reply:
Or perhpas we should have `globalState.allowStateReads`, which is true in action/derivation. That way the `untracked` wouldn't warn even with missing derivation.

Reply:
if we got to that point,
Can we chat about it directly and make an implementation plan?

Reply:
I've started putting together a list of what needs to be done, but there is something I came across:  
`checkIfStateModificationsAreAllowed` is called individually for each mutative operation per observable type (value/array/map):
https://github.com/mobxjs/mobx/search?q=checkIfStateModificationsAreAllowed&unscoped_q=checkIfStateModificationsAreAllowed 
I wonder why it isn't called always during `atom.reportChanged()` as it represents a *state change*...
It also means that if you create a custom atom, it doesn't respect the `enforceAction` by itself.  
My concern is where to place eventual `checkIfStateReadsAreAllowed` - whether it could be placed in `atom.reportObserved()`, or whether it has to be in `observablevalue.get()` and alike.

Reply:
reportChanged sounds like something after the action have happened,
And `checkIfStateModificationsAreAllowed` needs to be checked before that.

If we are going only to warn, maybe reportObserved can work

Reply:
Right, it would throw after the state was mutated. I don't think it would really matter in the end as I don't expect anyone will try recover from the exception, but it could definitely break some expectations.
I think it's impossible to obtain observable value without calling `reportObserved()` first and I can't imagine a scenario where you would want to call `reportObserved()` without warning, so let's stick with that for now.